design/external-references.md

```diff
@@ -0,0 +1,394 @@
+
+**Serialization context**: the notion of a "string key/value context" represents the lowest common denominator for serialization, and specifically is the only type supported by CloudFormation's cross-stack export/import mechanism (CloudFormation outputs can only be of "string" type). If a construct can serialize itself through a set of key/values, we can pass it via CloudFormation's import/export, environment variables, SSM parameters, etc.
+
+**CloudFormation imports/exports as a serialization context**: on top of that, we will implement a serialization/deserialization context based on AWS CloudFormation's import/exports mechanism. "writing" a value means defining an Output with an Export and "reading" a value means `Fn::ImportValue` with this export name. The "export name" will be used as a prefix that represent an object, and sub-objects will be serialized by adding another component to the export name.
+
+**Synthesis-time Imports**: we will also define an environmental context provider that will import values during synthesis and store them in `cdk.json`. This approach is more robust since it will allow CDK code to reason about concrete values instead of opaque tokens. It's especially needed for situations where a list of values needs to be passed and the arity is required.
+
+**Convenience methods**: now that constructs can be serializable through imports/exports, we can implement a set of convenience methods for each AWS resource to provide nice ergonomics for the specific use case of exporting and importing resources across apps.
+
+## Design
+
+### Serialization
+
+If a construct is said to be serializable, it must implement the `ISerializable` interface:
+
+```ts
+interface ISerializable {
+  serialize(ctx: ISerializationContext): void;
+}
+```
+
+When an object is serialized, the `serialize` method is called with an object which implements the following interface:
+
+```ts
+interface ISerializationContext {
+  writeString(key: string, value: string, options?: SerializationOptions): void
+  writeObject(key: string, obj: ISerializable, options?: SerializationOptions): void;
+}
+
+interface SerializationOptions {
+  description?: string;
+}
+```
+
+The serialization context allows the object serialize itself through key/value strings via calls to `writeString(key, value)`. If the object encapsulates another serializable object, it can use `writeObject(key, obj)`, which will result in a subsequent call to the sub-object's `serialize` method with an appropriate context.
+
+### Deserialization
+
+To support deserialization, classes must also include a public static `deserializeXxx` method which reads the object from a deserialization context and returns an object that implements the resource type interface:
+
+> The reason we indicate the type in the method name is because static methods in JavaScript are inherited, so we can differentiate between `ExtraBucket.deserializeExtraBucket` and `ExtraBucket.deserizlizeBucket`.
```

Comment:
IMO `ExtraBucket.deserializeBucket` does not make sense, but I get you cannot prevent it from being done (shame!). Maybe we should:

1. Make sure JSII does not model/understand statics as inheritable
2. Have a way to make sure the inherited version of a static cannot be called (is this even possible?)